src/rpm.rs

```diff
@@ -45,7 +54,71 @@ impl RPMPackage {
         out.write_all(&self.content)?;
         Ok(())
     }
+
+    /// sign all headers (except for the lead) using an external key and store it as the initial header
+    #[cfg(feature = "signing")]
+    pub fn sign(&mut self, secret_key: &[u8]) -> Result<(), RPMError> {
+        // create a temporary byte repr of the header
+        // and re-create all hashes
+        let mut header_bytes = Vec::<u8>::with_capacity(1024);
+        self.metadata.header.write(&mut header_bytes)?;
+
+        let _signature_size = header_bytes.len() + self.content.len();
+        let mut hasher = md5::Md5::default();
+
+        hasher.input(&header_bytes);
+        hasher.input(&self.content);
+
+        let hash_result = hasher.result();
+
+        let digest_md5 = hash_result.as_slice();
+
+        let digest_sha1 = sha1::Sha1::from(&header_bytes);
+        let digest_sha1 = digest_sha1.digest();
+
+        let signer = Signer::load_secret_key(secret_key)?;
+
+        // TODO verify the byte range is correct!
+        let signature_header = signer.sign(self.content.as_slice())?;
+        let _size = self.content.len();
+
+        let rsa_spanning_header_only = raw_signature_to_rfc4880(&signature_header[..])?;
+
+        let signature_header_and_archive = signer.sign(self.content.as_slice())?;
+        let size = self.content.len();
+
+        let rsa_spanning_header_and_archive =
+            raw_signature_to_rfc4880(&signature_header_and_archive[..])?;
+
+        // TODO FIXME verify this is the size we want, I don't think it is
+        // TODO maybe use signature_size instead of size
+        self.metadata.signature = Header::<IndexSignatureTag>::new_signature_header(
+            size as i32,
+            digest_md5,
+            digest_sha1.to_string(),
+            rsa_spanning_header_only.as_slice(),
+            rsa_spanning_header_and_archive.as_slice(),
+        );
+
+        Ok(())
+    }
+
+    /// verify the signature as present
+    #[cfg(feature = "signing")]
+    pub fn verify_signature(&mut self, public_key: &[u8]) -> Result<(), RPMError> {
+        // TODO retval should be SIGNATURE_VERIFIED or MISMATCH, not just an error
```

Comment:
I think we need to be able to tell the difference between "signature verification error" and "general rpm error" so that clients have a chance to act accordingly. Returning an empty result when everything is okay is fine by me.

Reply:
This can be done in two ways:
* treat signature verification as an error RPMError variant (current state)
* replace the `()` by e `enum ValidationOutcome { Passed, Failed}` and let the verifying party deal with it

The second option was favoured, but in the end the end it just causes additional mapping of `ValidationOutcome::Failed` back into an `RPMError` after all.

Reply:
tbh, current error handling in the whole crate is very basic and could use some sort of an overhaul. Let's just return the RPMError as is and I will think of a way how to make the error more usable for clients.